Update getting_started & sce_file_contents with celltyping instructions (#193)
AlexsLemonade/scpca-docs: docs/getting_started.md

 - [Quantifying clustering behavior in Orchestrating Single Cell Analysis](https://bioconductor.org/books/release/OSCA.advanced/clustering-redux.html#quantifying-clustering-behavior)

### Cell type annotation

Processed objects may contain cell type annotations and associated metadata from one or more of the following sources.

> - Submitter-provided annotations
>   - Note that these are only present for a subset of libraries.

Reviewer: ```suggestion
- Submitter-provided annotations (note that these are only present for a subset of libraries). 
```
I think you could just put this on the same line and don't need the nested bullet.

- Annotations determined by [`SingleR`](https://bioconductor.org/packages/release/bioc/html/SingleR.html), an automated reference-based method ([Looney _et al._ 2019](https://doi.org/10.1038/s41590-018-0276-y)).
- Annotations determined by [`CellAssign`](https://github.com/Irrationone/cellassign), an automated marker-gene based method ([Zhang _et al._ 2019](https://doi.org/10.1038/s41592-019-0529-1)).

> If at least one type of cell type annotation is present, a supplemental cell type report will be provided with the download.

Reviewer: ```suggestion
If at least one method was used for cell type annotation, a supplemental cell type report will be provided with the download.
```
I feel like we want to use the word "method" rather than "type" when referring to what kind of cell type annotations are present. I know submitter isn't really a method, but the other two are... So if you have a better word I'm open to other options!

Author: honestly i think we are doomed with this wording no matter what, so i'm ok with it!

This report evaluates cell annotations results as follows:

- It provides diagnostic plots to assess the quality of cell type annotations
- If multiple annotations are present, the report compares different annotations to one another.
Strong agreement between different annotation methods is a qualitative indicator of robustness.

> To determine which cell type annotations are present, use the following command on the processed `SingleCellExperiment` object:

Reviewer: ```suggestion
To determine which methods were used for cell type annotations, use the following command on the processed `SingleCellExperiment` object:
```

```r
# show vector of available celltypes
# values will be one or more of: `submitter`, `singler`, `cellassign`
metadata(processed_sce)$celltype_methods
```

Or, on the processed `AnnData` object:

```python
# show list of available celltypes
# values will be one or more of: `submitter`, `singler`, `cellassign`
processed_adata.uns["celltype_methods"]
```

> Below we provide instructions on how to access each type of cell type annotation, if present.

Reviewer: ```suggestion
Below we provide instructions on how to access annotations from each cell type annotation method used, if present.
```

```python
# submitter-provided annotations for each cell
processed_adata.obs["submitter_celltype_annotation"]
```

Cells that submitters did not annotate are labeled with `Submitter-excluded`.
Note that submitter-provided annotations are also present in unfiltered and filtered objects and can be accessed using the same approach shown here for processed objects.

#### `SingleR` annotations

`SingleR` annotation uses a reference dataset from the [`celldex` package](https://bioconductor.org/packages/release/data/experiment/html/celldex.html) [[Aran _et al._ (2019)](https://doi.org/10.1038/s41590-018-0276-y)].

To access automated `SingleR` annotations as cell type names and/or ontology terms in the process `SingleCellExperiment` object, use the following command(s):

```r
# SingleR annotatins as cell type names
processed_sce$singler_celltype_annotation

# Or, SingleR annotatins as cell type ontology terms
processed_sce$singler_celltype_ontology
```

To access automated `SingleR` annotations as cell type names and/or ontology terms in the processed `AnnData` object, use the following command(s):

> ```r

Reviewer: ```suggestion
```python
```

```

Cells that `SingleR` could not confidently annotate are labeled with `NA`.

You can also access the full object returned by `SingleR` from the `SingleCellExperiment`'s metadata with the following command (note that this information is not provided in the `AnnData` object):

```r
# SingleR full result
metadata(processed_sce)$singler_results
```

Additional information about `SingleR` annotation, including the reference source and version, is also available from the processed `SingleCellExperiment` object's metadata and from the processed `AnnData` object's `uns` slot, as described in the {ref}`experiment metadata table<sce_file_contents:SingleCellExperiment experiment metadata>`.

#### `CellAssign` annotations

`CellAssign` annotation uses a reference set of marker genes from the [`PanglaoDB` database](https://panglaodb.se/) [[Oscar Franzén _et al._ (2019)](https://doi.org/10.1093/database/baz046)], as compiled by the Data Lab for a given tissue group.

To access automated `CellAssign` annotations in the `SingleCellExperiment`, use the following command:

```r
# CellAssign annotations for each cell
processed_sce$cellassign_celltype_annotation
```

> Additional information about `CellAssign` annotation results are also available from the `SingleCellExperiment`'s metadata, as described in the {ref}`experiment metadata table<sce_file_contents:SingleCellExperiment experiment metadata>`.
> 

Reviewer: ```suggestion
```
I don't think you need this twice.

Author: I wonder if we should put 1 sentence about this, to cover both singler and cellassign, above the subsections where we go through each method. Something like, "in addition to what is described below, more information is available blah blah blah". What do you think?

Reviewer: I like that! I don't think you need that and this sentence that's at the end. I would just do one, but maybe repetition isn't bad when we are talking about user facing docs.

Author: > I don't think you need that and this sentence that's at the end.

I agree, I think just 1 at the top is ok! In this case, the repetition will probably make it all too long.

To access automated `CellAssign` annotations in the `AnnData` object, use the following command:

```python
# CellAssign annotations for each cell
processed_adata.obs["cellassign_celltype_annotation"]
```

Cells that `CellAssign` could not confidently annotate are labeled with `"other"`.

You can also access the full predictions matrix returned by `CellAssign` from the `SingleCellExperiment`'s metadata with the following command:

> ```r

Reviewer: add in the python command for this?

Author: Oh thanks for pointing this out, I had meant to ask about it. Are we sure we are saving this in AnnData? I don't see it in an object I was working with (which I generated last week) - 

```
>>> adata
AnnData object with n_obs × n_vars = 5047 × 60319
    obs: 'barcodes', 'submitter_celltype_annotation', 'sum', 'detected', 'subsets_mito_sum', 'subsets_mito_detected', 'subsets_mito_percent', 'total', 'prob_compromised', 'miQC_pass', 'scpca_filter', 'sizeFactor', 'cluster', 'singler_celltype_ontology', 'singler_celltype_annotation', 'cellassign_celltype_annotation', 'cellassign_max_prediction', 'library_id', 'sample_id', 'scpca_project_id', 'submitter_id', 'participant_id', 'submitter', 'age', 'sex', 'diagnosis', 'subdiagnosis', 'tissue_location', 'disease_timing', 'organism', 'development_stage_ontology_term_id', 'sex_ontology_term_id', 'organism_ontology_id', 'self_reported_ethnicity_ontology_term_id', 'disease_ontology_term_id', 'tissue_ontology_term_id', 'assay_ontology_term_id', 'suspension_type', 'is_primary_data'
    var: 'gene_ids', 'gene_symbol', 'mean', 'detected', 'feature_is_filtered'
    uns: 'X_name', 'af_num_cells', 'af_permit_type', 'af_resolution', 'alevinfry_version', 'assay_ontology_term_id', 'cellassign_reference', 'celltype_methods', 'cluster_algorithm', 'cluster_nn', 'cluster_weighting', 'filtering_method', 'highly_variable_genes', 'include_unspliced', 'library_id', 'mapped_reads', 'mapping_tool', 'min_gene_cutoff', 'normalization', 'prob_compromised_cutoff', 'project_id', 'reference_index', 'salmon_version', 'sample_id', 'schema_version', 'scpca_filter_method', 'seq_unit', 'singler_reference', 'singler_reference_label', 'tech_version', 'total_reads', 'transcript_type', 'usa_mode'
    obsm: 'X_PCA', 'X_UMAP'
    layers: 'logcounts', 'spliced'
```

But this seems to be the only place we removed content from the metadata - 

https://github.com/AlexsLemonade/scpca-nf/blob/263b6e7d94ae119323fd706efe5d1f157c577419/bin/sce_to_anndata.R#L89-L90

Reviewer: Ah you are right. I think we are removing it in the scpcaTools function. https://github.com/AlexsLemonade/scpcaTools/blob/bc767a55dbc0a702b89c73e6d6b5a8a45ba986f5/R/sce_to_anndata.R#L54

Reviewer: Ah you are right. I think we are removing it in the scpcaTools function. https://github.com/AlexsLemonade/scpcaTools/blob/bc767a55dbc0a702b89c73e6d6b5a8a45ba986f5/R/sce_to_anndata.R#L54

 - [Quantifying clustering behavior in Orchestrating Single Cell Analysis](https://bioconductor.org/books/release/OSCA.advanced/clustering-redux.html#quantifying-clustering-behavior)

> ### Cell type annotation

Reviewer: Generally, I think we want to add a little bit more context to this section. I would describe the references used in this first section. I would also mention the supplemental cell type report and that it's a good resource for evaluating the provided cell type annotations.

Reviewer: I think what you added is good!

Processed objects may contain cell type annotations and associated metadata from one or more of the following sources.

- Submitter-provided annotations

> - Automated annotations from [`SingleR`](https://bioconductor.org/packages/release/bioc/html/SingleR.html), a reference-based method ([Looney _et al._ 2019](https://doi.org/10.1038/s41590-018-0276-y)).
> - Automated annotations from [`CellAssign`](https://github.com/Irrationone/cellassign), a marker-gene based method ([Zhang _et al._ 2019](https://doi.org/10.1038/s41592-019-0529-1)).

Reviewer: ```suggestion
- Annotations determined by using [`SingleR`](https://bioconductor.org/packages/release/bioc/html/SingleR.html), a reference-based method ([Looney _et al._ 2019](https://doi.org/10.1038/s41590-018-0276-y)).
- Annotations determined by using [`CellAssign`](https://github.com/Irrationone/cellassign), a marker-gene based method ([Zhang _et al._ 2019](https://doi.org/10.1038/s41592-019-0529-1)).
```
I don't know if this is much better, but we didn't take annotations from these methods, we performed annotation using those methods.

Author: I think this could be a good middle ground - `Annotations determined by <method>, an automated <thing-based> method`

Below we provide instructions on how to access each type of cell type annotation, if present.

#### Submitter-provided annotations

> To access submitter-provided annotations in the `SingleCellExperiment`, use the following command:

Reviewer: I think we should add a note that not all libraries will contain submitter annotations. Although everything should include CellAssign/SingleR, it's much more likely that these will be missing.

```r
# submitter-provided annotations for each cell
processed_sce$submitter_celltype_annotation
```

To access submitter-provided annotations in the `AnnData` object, use the following command:

```python
# submitter-provided annotations for each cell
processed_adata.obs["submitter_celltype_annotation"]
```

> Cells that submitters did not annotate are labeled with `NA`.

Reviewer: ```suggestion
Cells that submitters did not annotate are labeled with `submitter-excluded`.
```
Didn't we decide on this?

Author: Ahhh yes, I forgot to update this after we decided!

Note that submitter-provided annotations are also present in unfiltered and filtered objects and can be accessed using the same approach shown here for processed objects.

#### `SingleR` annotations

To access automated `SingleR` annotations in the `SingleCellExperiment`, use the following command:

```r
# SingleR annotations for each cell
processed_sce$singler_celltype_annotation
```

To access automated `SingleR` annotations in the `AnnData` object, use the following command:

```python
# SingleR annotations for each cell
processed_adata.obs["singler_celltype_annotation"]
```

Cells that `SingleR` could not confidently annotate are labeled with `NA`.

If `SingleR` annotation was performed using ontology reference labels (default), the cell type annotation ontology terms will also be provided.

To access cell type annotation ontology terms in the `SingleCellExperiment` object, use the following command:
```r
# SingleR ontology annotations
processed_sce$singler_celltype_ontology
```

To access cell type annotation ontology terms in the `AnnData` object, use the following command:
```r
# SingleR ontology annotations
processed_adata.obs["singler_celltype_ontology"]
```

> 

Reviewer: Since we are always using the ontology label for the portal, I would condense this with the previous commands. And say that we are storing both the annotations and ontology labels and then in the example code chunk include both lines of code.

```r
# SingleR full result
metadata(processed_sce)$singler_results
```

`SingleR` annotation uses a reference dataset from the [`celldex` package](https://bioconductor.org/packages/release/data/experiment/html/celldex.html) [[Aran _et al._ (2019)](https://doi.org/10.1038/s41590-018-0276-y)].

To access information about the `SingleR` reference in the `SingleCellExperiment` object, use the following commands:

```r
# SingleR reference dataset name
metadata(processed_sce)$singler_reference

# SingleR reference dataset label used for annotation
metadata(processed_sce)$singler_reference_label

# SingleR reference version
metadata(processed_sce)$singler_reference_version
```

To access the analogous information in the `AnnData` object, use the following commands:

```python
# SingleR reference (celldex) dataset name
processed_adata.uns["singler_reference"]

# SingleR reference (celldex) dataset label used for annotation
processed_adata.uns["singler_reference_label"]

# SingleR reference (celldex) version
processed_adata.uns["singler_reference_version"]
```

<!-- TODO: this instead? -->

> Additional information about `SingleR` annotation results is also available from the `SingleCellExperiment`'s metadata, as described in the {ref}`experiment metadata table<sce_file_contents:SingleCellExperiment experiment metadata>`.

Reviewer: Personally, I think this is a little long and we don't give this much detail about metadata in the other sections. I think you replace the version information with a link like you suggest here.

Author: I think I'll also be a little more explicit like `Additional information, including reference and version metadata, ...`

You can also access the full predictions matrix returned by `CellAssign` from the `SingleCellExperiment`'s metadata with the following command:

```r
# CellAssign full predictions matrix full result
metadata(processed_sce)$cellassign_predictions
```

`CellAssign` annotation uses a reference set of marker genes from the [`PanglaoDB` database](https://panglaodb.se/) [[Oscar Franzén _et al._ (2019)](https://doi.org/10.1093/database/baz046)], as compiled by the Data Lab for a given tissue group.

To access information about the `CellAssign` reference in the `SingleCellExperiment` object, use the following commands:

```r
# CellAssign reference name as labeled by the Data Lab
metadata(processed_sce)$cellassign_reference

# CellAssign reference (PanglaoDB) version
metadata(processed_sce)$cellassign_reference_version
```

To access the analogous information in the `AnnData` object, use the following commands:

```python
# CellAssign reference name as labeled by the Data Lab
processed_adata.uns["cellassign_reference"]

# CellAssign reference (PanglaoDB) version
processed_adata.uns["cellassign_reference_version"]
```

> <!-- TODO: this instead? -->

Reviewer: Same comment here as above.
